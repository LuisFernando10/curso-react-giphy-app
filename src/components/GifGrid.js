    
    import React from 'react';
    import PropTypes from 'prop-types';
    import { useFetchGifs } from '../hooks/useFetchGifs';
    import { GifGridItem } from './GifGridItem';

    export const GifGrid = ( { category } ) => {

        /*const [images, setImages] = useState( [] );

        /// El Hook 'UseEffect', permite ejecutar una única vez cuando el componente es renderizado por primera vez
        useEffect( () => {
            getGifs( category )
                .then( setImages );// Sólo se coloca la referencia al el método porque la promesa y el método reciben un parámetro, entonces se omite
        }, [ category ]);// Agregamos la 'categoría' para que si 'cambia', entonces ejecutamos nuevamente el Effect
        */

        const { data:images, loadig } = useFetchGifs( category ); // 'data:images', es la forma de renombrar una 'prop' cuando en el archivo local, queremos personalizar su nombre

        return (
            <>
                <h2>{ category }</h2>

                {/* { loadig ? <p>Cargando...</p> : null } //Ésta es la forma larga de evaluar dos condiciones  */}
                { loadig && <p className='animate__animated animate__shakeX'>Cargando...</p> }
                <div className='card-grid'>
                    {
                        images.map( ( img ) => (
                            <GifGridItem
                                key={ img.id }
                                { ...img }
                            />
                        ))
                    }
                </div>
            </>
        )
    }

    GifGrid.propTypes = {
        category: PropTypes.string.isRequired
    }