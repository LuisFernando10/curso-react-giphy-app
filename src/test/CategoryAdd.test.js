
    import React from 'react';
    import '@testing-library/jest-dom'; // Para tener las ayudas
    import { shallow } from 'enzyme';
    import { CategoryAdd } from '../components/CategoryAdd';

    describe('Prueba en CategoryAdd.js', () => { 
        
        const setCategories = jest.fn(); // Jest crea una función por default para usar como referencia
        let wrapper = shallow(<CategoryAdd setCategories={ setCategories } />);

        // Reinicio de procesos/datos (antes de cada prueba)
        beforeEach( () => {
            jest.clearAllMocks(); // Limpiamos simulaciones o mocks/funciones en jest
            wrapper = shallow(<CategoryAdd setCategories={ setCategories } />);
        });

        test('Debe: Retornar componente < CategoryAdd />', () => { 
            expect( wrapper ).toMatchSnapshot();
        });

        test('Debe: Detectar cambio en la caja de texto', () => {

            const element_input = wrapper.find('input');
            const value = 'Hola Mundo';

            element_input.simulate('change', {
                target: {
                    value
                }
            });

            expect( wrapper.find('p').text().trim() ).toBe( value );
        });

        test('Debe: Bloquear el submit cuando no hay valor en el input', () => { 
          
            //wrapper.find('form').simulate('submit', { preventDefault: () => {} }); // Forma larga para función del 'preventDefault'
            wrapper.find('form').simulate('submit', { preventDefault(){} }); // Forma corta de definir una referencia de función para el 'preventDefault'

            expect( setCategories ).not.toHaveBeenCalled();// Método para saber si la función 'setCategories' se ejecutó 1 o más veces
        });

        test('Debe: llamar setCateories y limpiar caja de texto', () => {
          
            const value = 'Hola Mundo';

            // 1. (Simular el Change), cambiando el valor del input y estableciendo un valor 'const value'
            wrapper.find('input').simulate('change', { target: { value } });

            // 2. (Simular el Submit)
            wrapper.find('form').simulate('submit', { preventDefault(){} });

            // 3. Llamar a 'setCategory' por lo menos una vez
            expect( setCategories ).toHaveBeenCalled(); // Llamada al menos 1 vez
            // expect( setCategories ).toHaveBeenCalledTimes(2); // Llamada por lo menos 2 veces 
            // expect( setCategories ).toHaveBeenCalledWith( expect.any(function) ); // Valida que el argumento sea estrictamente cualquier tipo de función

            // 4. El valor del input se debe setear a: ''
            expect( wrapper.find('input').prop('value') ).toBe('');
        });
     });