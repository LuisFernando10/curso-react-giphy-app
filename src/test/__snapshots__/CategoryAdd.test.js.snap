// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en CategoryAdd.js Debe: Retornar componente < CategoryAdd /> 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
